rust/chains/hyperlane-ethereum/src/merkle_tree_hook.rs

```diff
@@ -0,0 +1,282 @@
+where
+    M: Middleware,
+{
+    contract: Arc<MerkleTreeHookContract<M>>,
+    domain: HyperlaneDomain,
+    provider: Arc<M>,
+}
+
+impl<M> EthereumMerkleTreeHook<M>
+where
+    M: Middleware,
+{
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(provider: Arc<M>, locator: &ContractLocator) -> Self {
+        Self {
+            contract: Arc::new(MerkleTreeHookContract::new(
+                locator.address,
+                provider.clone(),
+            )),
+            domain: locator.domain.clone(),
+            provider,
+        }
+    }
+}
+
+impl<M> HyperlaneChain for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(EthereumProvider::new(
+            self.provider.clone(),
+            self.domain.clone(),
+        ))
+    }
+}
+
+impl<M> HyperlaneContract for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    fn address(&self) -> H256 {
+        self.contract.address().into()
+    }
+}
+
+#[async_trait]
+impl<M> MerkleTreeHook for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    #[instrument(skip(self))]
+    async fn latest_checkpoint(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<Checkpoint> {
+        let lag = maybe_lag.map(|v| v.get()).unwrap_or(0).into();
+
+        let fixed_block_number: BlockNumber = self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .saturating_sub(lag)
+            .into();
+
+        let (root, index) = self
+            .contract
+            .latest_checkpoint()
+            .block(fixed_block_number)
+            .call()
+            .await?;
+        Ok(Checkpoint {
+            merkle_tree_hook_address: self.address(),
+            mailbox_domain: self.domain.id(),
+            root: root.into(),
+            index,
+        })
+    }
+
+    #[instrument(skip(self))]
+    #[allow(clippy::needless_range_loop)]
+    async fn tree(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<IncrementalMerkle> {
+        let lag = maybe_lag.map(|v| v.get()).unwrap_or(0).into();
+
+        let fixed_block_number: BlockNumber = self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .saturating_sub(lag)
+            .into();
+
+        // TODO: implement From<Tree> for IncrementalMerkle
+        let raw_tree = self
+            .contract
+            .tree()
+            .block(fixed_block_number)
+            .call()
+            .await?;
+        let branch = raw_tree
+            .branch
+            .iter()
+            .map(|v| v.into())
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
```

Comment:
should we be unwrapping? dunno if this was there before - if it was, worth checking if we can easily remove, but if we can't then nbd

Reply:
the unwrap is safe here. It's required because although we're iterating over elements of a fixed-size array (`[u8; 32]`), `collect` can only produce a collection that implements `FromIterator` - which isn't the case for fixed-size arrays because an iterator has no knowledge about how many items it's iterating over. So we need the `try_into()` to convert the `Vec<H256>` into `[H256; 32]`. Alternative would be
```rust
        let mut branch: [H256; 32] = [Default::default(); 32];
        raw_tree
            .branch
            .iter()
            .enumerate()
            .for_each(|(index, v)| branch[index] = v.into());
```
This code can still panic with an out-of-bounds access if `raw_tree.branch` exceeds 32, and would probably be less obvious to debug than with the unwrap version. I'd prefer leaving as-is and adding a comment, what's your take?